### PR TITLE
[WIP] tor V2/V3 connect support for lightning

### PR DESCRIPTION
**Pull request by Saibato**

### CODE REVIEW

**Comment by Saibato** [comment]
This a just a [WIP] but functional the addrlen for V2/V2 must be for sure decoded/encode  32/64 to match BOLT spec.
can we use external libs?
Still some part CI wont like pintf whitespace etc.

**Comment by ZmnSCPxj** [comment]
We should prefer to reduce our external dependencies; my understanding is that the base32 encoding should be reasonably simple enough that it can be implemented within a few hours.  In particular we would want to make the base32 translation use the `tal` system we use for the rest of the code.

**Comment by Saibato** [comment]
OK  I found 2 short sources for the 64/32 the decode i will rewrite it to fit with tal_

**Comment by ZmnSCPxj** [comment]
We should not be using `printf`, we should use `log_*` functions in `lightningd/`, or `status_*` in the subdaemons.  `strlen` is OK to use, in general we use standard strings, we just allocate them using `tal` system.

**Comment by Saibato** [comment]
base 32/64 ready make some test  final max addrsize 42 :-) will go for bugfixing and try to crash the progiee have some spare time this week

**Comment by Saibato** [comment]
@ZmnSCPxj  moin moin , can you please review?  Its fully functional but CI still lammers
And help me out with some special port and V3 tests ? 
a node with ID  02295947d90891435be4bf16735854565aa7084c45ea4ab462f8d3ecebc35e9cbc  
and tor v3 addr
4ruvswpqec5i2gogopxl4vm5bruzknbvbylov2awbo4rxiq4cimdldad.onion:49150 it will be online for 1 hour 
i need only one ping ^^
Now c-LN can easy hop over FW next step will be to make it a full auto service by --torlisten option then things can be transfered to smart contracts and stay there for ever.

File: common/wireaddr.c [hunk 1]
```diff
@@ -56,6 +231,12 @@ char *fmt_wireaddr(const tal_t *ctx, const struct wireaddr *a)
 		if (!inet_ntop(AF_INET6, a->addr, addrstr, INET6_ADDRSTRLEN))
 			return "Unprintable-ipv6-address";
 		return tal_fmt(ctx, "[%s]:%u", addrstr, a->port);
+    case ADDR_TYPE_TOR_V2:
+    
+		return tal_fmt(ctx, "%s.onion:%u", b32_encode(addrstr, (u8 *)a->addr) , a->port);  
+	case ADDR_TYPE_TOR_V3:
+		return tal_fmt(ctx, "%s.onion:%u", b64_encode(a->addr,42) , a->port);
```
Comment (ZmnSCPxj) [inline common/wireaddr.c]: Should it not be using `b32_encode`? https://gitweb.torproject.org/torspec.git/tree/proposals/224-rend-spec-ng.txt#n2003 seems to say that the new V3 addresses still use base32 (the logic being that it replaces the domain name, and domain names are traditionally case insensitive, and base64 would introduce case sensitivity).  I believe we do not need base64 at all.

File: common/base64.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+#ifndef B64_H
+#define B64_H 1
+
+/**
+ *  Memory allocation functions to use. You can define b64_malloc and
+ * b64_realloc to custom functions if you want.
+ */
+
+#ifndef b64_malloc
```
Comment (ZmnSCPxj) [inline common/base64.h]: We should be using only `tal` and related functions.  So things like `b32_encode` should accept a `const tal_t *ctx` argument that will own the returned encoded string.

File: common/wireaddr.c [hunk 3]
```diff
@@ -136,14 +437,14 @@ bool parse_wireaddr(const char *arg, struct wireaddr *addr, u16 defport,
...
 	if (streq(ip, "localhost"))
 		ip = "127.0.0.1";
 	else if (streq(ip, "ip6-localhost"))
 		ip = "::1";
 
-	memset(&addr->addr, 0, sizeof(addr->addr));
-
-	if (inet_pton(AF_INET, ip, &v4) == 1) {
+    memset(&addr->addr, 0, sizeof(addr->addr)); 
+if (inet_pton(AF_INET, ip, &v4) == 1) {
```
Comment (ZmnSCPxj) [inline common/wireaddr.c]: Unnecessary change?

File: common/wireaddr.c [hunk 4]
```diff
@@ -10,6 +10,174 @@
...
+int b32_decode( u8 *dst,u8 *src) {
+ 
+  int rem = 0;
+ 
+  int i;
+  u8 *p=src;
+  int buf;
+  u8 ch;
+  
+  for (i=0; i < 16 ; p++) {
```
Comment (ZmnSCPxj) [inline common/wireaddr.c]: Truly fixed size?  `p` is incremented here but the loop variable is `i`.  You do increment `i` later in the code but maybe better to put the `p` increment in the loop body and the `i` increment here.

File: common/wireaddr.c (hunk 1 above)
Comment (Saibato) [reply common/wireaddr.c]: removed all base64 changed to base32

File: common/base64.h (hunk 2 above)
Comment (Saibato) [reply common/base64.h]: wip

File: common/wireaddr.c (hunk 3 above)
Comment (Saibato) [reply common/wireaddr.c]: yepp

File: common/wireaddr.c (hunk 4 above)
Comment (Saibato) [reply common/wireaddr.c]: this is tricky the magic is the increment of (i)  on dep of bits > 111..... 
so (i) i cant be in the main loop but the next char has to be fetched.

**Comment by Saibato** [comment]
@ZmnSCPxj  have done some cleanups function test pass 
but CI (vallgrinds) does lammer   
allways on
DEBUG:root:Received response for listfunds call: {'jsonrpc': '2.0', 'result': {'channels': [], 'outputs': [{'output': 0, 'value': 10000000, 'txid': '08125e35542654949b15daba28c5caaf2659bc624ee4acc2d514a7463ca194b5'}]}, 'id': 0}
DEBUG:root:Calling stop with payload None
DEBUG:root:lightningd(16331): 2018-03-01T21:50:18.550Z lightningd(1604): Connected json input
DEBUG:root:lightningd(16331): 2018-03-01T21:50:18.551Z lightningd(1604):jcon fd 16: Success
DEBUG:root:Received response for stop call: {'jsonrpc': '2.0', 'result': 'Shutting down', 'id': 0}
DEBUG:root:lightningd(16331): 2018-03-01T21:50:18.552Z lightningd(1604):jcon fd 16: JSON-RPC shutdown
DEBUG:root:lightningd(16331): 2018-03-01T21:50:18.554Z lightning_gossipd(1613): Shutting down
DEBUG:root:bitcoind: 2018-03-01 21:50:19 Shutdown: In progress...
DEBUG:root:bitcoind: 2018-03-01 21:50:19 addcon thread exit
DEBUG:root:bitcoind: 2018-03-01 21:50:19 opencon thread exit
DEBUG:root:bitcoind: 2018-03-01 21:50:19 net thread exit
DEBUG:root:bitcoind: 2018-03-01 21:50:19 msghand thread exit
DEBUG:root:bitcoind: 2018-03-01 21:50:19 scheduler thread interrupt
DEBUG:root:bitcoind: 2018-03-01 21:50:19 Dumped mempool: 2e-06s to copy, 0.002042s to dump
DEBUG:root:bitcoind: 2018-03-01 21:50:19 Shutdown: done
------------------------------- Valgrind errors --------------------------------
Valgrind error file: valgrind-errors.1604
==1604== Conditional jump or move depends on uninitialised value(s)
==1604==    at 0x46E4D6: tal_free (tal.c:501)
==1604==    by 0x41B45E: setup_default_config (options.c:476)
==1604==    by 0x41BED9: handle_opts (options.c:672)
==1604==    by 0x412E72: main (lightningd.c:311)
==1604== 
==1604== Conditional jump or move depends on uninitialised value(s)
==1604==    at 0x4131D3: main (lightningd.c:387)
==1

at 476
is
/* Set default PID file name to be per-network */
	tal_free(ld->pidfile);
	ld->pidfile = tal_fmt(ld, "lightningd-%s.pid", get_chainparams(ld)->network_name);

but i dont touched this is original master, any ideas ???

File: common/wireaddr.c [hunk 5]
```diff
@@ -12,137 +12,33 @@
...
+#define BASE32DATA "abcdefghijklmnopqrstuvwxyz234567"
 
-    // while there is still a remainder
-    // append `=' to `enc'
-    while ((i++ < 3)) {
-      enc = (char *) b64_realloc(enc, size + 1);
-      enc[size++] = '=';
-    }
-  }
+char *b32_encode(char *, u8 *,u8);
```
Comment (ZmnSCPxj) [inline common/wireaddr.c]: The function is defined and used in this file only.  Make it `static`, also, remove the declaration, just define it directly.

File: common/wireaddr.c [hunk 6]
```diff
@@ -12,137 +12,33 @@
...
-}
-
-
-#define BASE32_DATA "abcdefghijklmnopqrstuvwxyz234567"
-
-char *b32_encode(char *, u8 *) ;
-
-
-char *b32_encode(char *dst, u8 *src) {
+char *b32_encode(char *dst, u8 *src, u8 ver) {
```
Comment (ZmnSCPxj) [inline common/wireaddr.c]: Comment describing the function would be good.  Does it require `tal_arr` buffers or not? (it does not).  Indicate that it returns `dst`.  And so on.

Suggest replacing `ver` with `size_t srclen`.

File: common/wireaddr.c [hunk 7]
```diff
@@ -459,41 +242,36 @@ if (inet_pton(AF_INET, ip, &v4) == 1) {
...
     };
        
-
-
 	/* Resolve with getaddrinfo */
 	if (!res) {
 		memset(&hints, 0, sizeof(hints));
-
-		hints.ai_family = AF_UNSPEC;
+        hints.ai_family = AF_UNSPEC;
```
Comment (ZmnSCPxj) [inline common/wireaddr.c]: Unnecessary change?

File: gossipd/gossip.c [hunk 8]
```diff
@@ -50,7 +50,6 @@
 #include <wire/wire_sync.h>
 
 #include <common/socksd.h>
```
Comment (ZmnSCPxj) [inline gossipd/gossip.c]: Include files need to be sorted.

**Comment by cdecker** [comment]
@Saibato why was this closed? I thought this was coming along great.
